README.md
rust-sysfs-led
==============

Bindings to Linux sysfs interface for LEDs.


## License

```
Copyright (c) 2017, Nick Stevens <nick@bitcurry.com>

Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
http://www.apache.org/license/LICENSE-2.0> or the MIT license
<LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
option.  This file may not be copied, modified, or distributed
except according to those terms.
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

### Added
- Initial release


src/colors.rs
// Copyright (c) 2017 Nick Stevens <nick@bitcurry.com>

//! Colorspace tools for RGB LEDs

use std::cmp;

/// RGB Black
pub const BLACK: Color = Color(0, 0, 0);
/// RGB White
pub const WHITE: Color = Color(255, 255, 255);
/// RGB Red
pub const RED: Color = Color(255, 0, 0);
/// RGB Green
pub const GREEN: Color = Color(0, 255, 0);
/// RGB Blue
pub const BLUE: Color = Color(0, 0, 255);
/// RGB Yellow
pub const YELLOW: Color = Color(255, 255, 0);
/// RGB Cyan
pub const CYAN: Color = Color(0, 255, 255);
/// RGB Magenta
pub const MAGENTA: Color = Color(255, 0, 255);

/// Representation of color in RGB colorspace
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub struct Color(u8, u8, u8);

impl Color {
    /// Create a new `Color` from red, green, and blue components
    pub fn from_rgb(red: u8, green: u8, blue: u8) -> Color {
        Color(red, green, blue)
    }

    /// Create a new `Color` from hue, saturation, and value components.
    ///
    /// Create a `Color` from HSV. Hue is the angle on a circle, with 0 equal
    /// to 0 degrees and 255 equal to 360 degrees. Saturation and Value are
    /// percents, with 0 equal to 0%, and 255 equal to 100%.
    pub fn from_hsv(hue: u8, saturation: u8, value: u8) -> Color {
        if saturation == 0 {
            // color is greyscale
            return Color(value, value, value);
        }

        // make hue 0-5
        let region = hue / 43;
        // find remainder part, make it from 0-255
        let fpart = (hue % 43) * 6;

        // calculate temp vars, doing integer multiplication
        let f = fpart as u16;
        let v = value as u16;
        let s = saturation as u16;
        let p = ((v * (255 - s)) >> 8) as u8;
        let q = ((v * (255 - ((s * f) >> 8))) >> 8) as u8;
        let t = ((v * (255 - ((s * (255 - f)) >> 8))) >> 8) as u8;

        match region {
            0 => Color(value, t, p),
            1 => Color(q, value, p),
            2 => Color(p, value, t),
            3 => Color(p, q, value),
            4 => Color(t, p, value),
            _ => Color(value, p, q),
        }
    }

    /// Create a new `Color` from hue, saturation, and lightness components.
    ///
    /// Create a `Color` from HSL. Hue is the angle on a circle, with 0 equal
    /// to 0 degrees and 255 equal to 360 degrees. Saturation and Lightness are
    /// percents, with 0 equal to 0%, and 255 equal to 100%.
    pub fn from_hsl(hue: u8, saturation: u8, lightness: u8) -> Color {
        if saturation == 0 {
            // color is greyscale
            return Color(lightness, lightness, lightness);
        }

        // make hue 0-5
        let region = hue / 43;
        // find remainder part, make it from 0-255
        let fpart = (hue % 43) * 6;

        // calculate temp vars, doing integer multiplication
        let f = fpart as u16;
        let l = lightness as u16;
        let s = saturation as u16;

        let chroma = if saturation < 128 {
            s.saturating_mul(l) >> 7
        } else {
            s.saturating_mul(255 - l) >> 7
        };

        let m = l.saturating_sub(chroma >> 1) as u8;
        let c = (chroma as u8).saturating_add(m);
        let x1 = ((chroma.saturating_mul(f) >> 8) as u8).saturating_add(m);
        let x2 = ((chroma.saturating_mul(255 - f) >> 8) as u8).saturating_add(m);

        match region {
            0 => Color(c, x1, m),
            1 => Color(x2, c, m),
            2 => Color(m, c, x1),
            3 => Color(m, x2, c),
            4 => Color(x1, m, c),
            _ => Color(c, m, x2),
        }
    }

    // pub fn to_hsl(&self) -> (u8, u8, u8) {
    //     let red = self.red() as u16;
    //     let green = self.green() as u16;
    //     let blue = self.blue() as u16;

    //     let cmax = cmp::max(cmp::max(red, green), blue);
    //     let cmin = cmp::min(cmp::min(red, green), blue);
    //     let delta = cmax - cmin;

    //     let hue = if delta == 0 {
    //         0
    //     } else if cmax == red {
    //         43 * (green - blue) / delta
    //         43 * (((self.green() - self.blue()) / delta) % 6)
    //     } else if cmax == self.green() {
    //         43 * (((self.blue() - self.red()) / delta ) + 2)
    //     } else {
    //         43 * (((self.red() - self.green()) / delta) + 4)
    //     };

    //     let lightness = (cmax + cmin) >> 1;

    //     let saturation = if delta == 0 {
    //         0
    //     } else if lightness < 128 {
    //         delta / (255 - (lightness << 1))
    //     } else {
    //         delta / ((lightness << 1) - 255)
    //     };

    //     (hue, saturation, lightness)
    // }

    pub fn red(&self) -> u8 {
        self.0
    }

    pub fn green(&self) -> u8 {
        self.1
    }

    pub fn blue(&self) -> u8 {
        self.2
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hsv_to_rgb() {
        assert_eq!(Color(0, 0, 0), Color::from_hsv(0, 0, 0));
        assert_eq!(Color(255, 255, 255), Color::from_hsv(0, 0, 255));
        assert_eq!(Color(255, 0, 0), Color::from_hsv(0, 255, 255));
        assert_eq!(Color(0, 255, 0), Color::from_hsv(86, 255, 255));
        assert_eq!(Color(0, 0, 255), Color::from_hsv(172, 255, 255));
        assert_eq!(Color(254, 255, 0), Color::from_hsv(43, 255, 255));
        assert_eq!(Color(0, 254, 255), Color::from_hsv(129, 255, 255));
        assert_eq!(Color(255, 0, 254), Color::from_hsv(215, 255, 255));
        assert_eq!(Color(192, 192, 192), Color::from_hsv(0, 0, 192));
        assert_eq!(Color(128, 128, 128), Color::from_hsv(0, 0, 128));
        assert_eq!(Color(128, 0, 0), Color::from_hsv(0, 255, 128));
        assert_eq!(Color(127, 128, 0), Color::from_hsv(43, 255, 128));
        assert_eq!(Color(0, 128, 0), Color::from_hsv(86, 255, 128));
        assert_eq!(Color(128, 0, 127), Color::from_hsv(215, 255, 128));
        assert_eq!(Color(0, 128, 126), Color::from_hsv(128, 255, 128));
        assert_eq!(Color(0, 0, 128), Color::from_hsv(172, 255, 128));
    }

    #[test]
    fn test_hsl_to_rgb() {
        assert_eq!(Color(  0,   0,   0), Color::from_hsl(  0,   0,   0));
        assert_eq!(Color(255, 255, 255), Color::from_hsl(  0,   0, 255));
        assert_eq!(Color(255, 255, 255), Color::from_hsl(255,   0, 255));
        assert_eq!(Color(255, 255, 255), Color::from_hsl(255, 255, 255));
        assert_eq!(Color(127, 127, 127), Color::from_hsl(  0,   0, 127));
        assert_eq!(Color(255,   0,   0), Color::from_hsl(  0, 255, 127));
        assert_eq!(Color(255, 125,   0), Color::from_hsl( 21, 255, 127));
        assert_eq!(Color(254, 255,   0), Color::from_hsl( 43, 255, 127));
        assert_eq!(Color(128, 255,   0), Color::from_hsl( 64, 255, 127));
        assert_eq!(Color(  0, 255, 251), Color::from_hsl(128, 255, 127));
        assert_eq!(Color(125,   0, 255), Color::from_hsl(193, 255, 127));
        assert_eq!(Color(190, 126,  64), Color::from_hsl( 21, 127, 127));
        assert_eq!(Color(189, 190,  64), Color::from_hsl( 43, 127, 127));
        assert_eq!(Color(127, 190,  64), Color::from_hsl( 64, 127, 127));
        assert_eq!(Color( 64, 190, 188), Color::from_hsl(128, 127, 127));
        assert_eq!(Color(126,  64, 190), Color::from_hsl(193, 127, 127));
    }
}


src/triggers.rs
// Copyright (c) 2017 Nick Stevens <nick@bitcurry.com>

use errors::*;
use super::{SysfsLed, SysfsRgbLed};

pub trait TriggerNone {
    fn none(&mut self) -> Result<()>;
}

impl TriggerNone for SysfsLed {
    fn none(&mut self) -> Result<()> {
        self.sysfs_write_file("trigger", "none")
    }
}

impl TriggerNone for SysfsRgbLed {
    fn none(&mut self) -> Result<()> {
        self.red.sysfs_write_file("trigger", "none")
            .and(self.green.sysfs_write_file("trigger", "none"))
            .and(self.blue.sysfs_write_file("trigger", "none"))
    }
}

pub trait TriggerTimer {
    fn timer(&mut self, delay_on: u64, delay_off: u64) -> Result<()>;
}

impl TriggerTimer for SysfsLed {
    fn timer(&mut self, delay_on: u64, delay_off: u64) -> Result<()> {
        self.sysfs_write_file("trigger", "timer")
            .and(self.sysfs_write_file("delay_on", &format!("{}", delay_on)))
            .and(self.sysfs_write_file("delay_off", &format!("{}", delay_off)))
    }
}

pub trait TriggerHeartbeat {
    fn heartbeat(&mut self, invert: bool) -> Result<()>;
}

impl TriggerHeartbeat for SysfsLed {
    fn heartbeat(&mut self, invert: bool) -> Result<()> {
        self.sysfs_write_file("trigger", "heartbeat")
            .and(self.sysfs_write_file("invert", if invert { "1" } else { "0" }))
    }
}

pub trait TriggerCpu {
    fn cpu(&mut self, cpu: u32) -> Result<()>;
}

impl TriggerCpu for SysfsLed {
    fn cpu(&mut self, cpu: u32) -> Result<()> {
        self.sysfs_write_file("trigger", &format!("cpu{}", cpu))
    }
}


examples/all_feature_demo.rs
#[macro_use]
extern crate error_chain;
extern crate sysfs_led;

use std::thread;
use std::time::Duration;
use sysfs_led::{Brightness, Led, RgbLed, SysfsLed, SysfsRgbLed};
use sysfs_led::colors::Color;
use sysfs_led::errors::*;

quick_main!(run);

fn run() -> Result<()> {
    let mut red = SysfsLed::new("redLed")?;
    let mut green = SysfsLed::new("grnLed")?;
    let mut blue = SysfsLed::new("bluLed")?;

    println!("Clear LEDs");
    red.set_brightness(Brightness::Off)?;
    green.set_brightness(Brightness::Off)?;
    blue.set_brightness(Brightness::Off)?;

    println!("Red LED on");
    red.set_brightness(Brightness::Full)?;
    delay();
    red.set_brightness(Brightness::Off)?;

    println!("Green LED on");
    green.set_brightness(Brightness::Full)?;
    delay();
    green.set_brightness(Brightness::Off)?;

    println!("Blue LED on");
    blue.set_brightness(Brightness::Full)?;
    delay();
    blue.set_brightness(Brightness::Off)?;

    println!("Flash Red LED");

    // let mut led = SysfsRgbLed::new("redLed", "grnLed", "bluLed").unwrap();
    // loop {
    //     for i in 0..255u8 {
    //         led.set_color(Color::from_hsl(i, 128, 32)).unwrap();
    //     }
    // }

    Ok(())
}

fn delay() {
    thread::sleep(Duration::new(3, 0));
}


src/lib.rs
// Copyright (c) 2017 Nick Stevens <nick@bitcurry.com>

//! Rust bindings for the [Linux sysfs LED class].
//!
//! Methods and types in this crate center around the [`Led`]
//! and [`RgbLed`] types. Implementations of these types are
//! provided for the Linux sysfs LED API, although an attempt has been made to
//! keep types generic enough that non-sysfs implementions could be possible.
//!
//! In addition to providing safe access to LED types directly, this crate also
//! provides access to a number of `Trigger`s through the [`triggers`] module.
//! Applying a trigger to an [`Led`] or [`RgbLed`] type can introduce a number
//! of automated behaviors, such as [timing],
//! [heartbeat], or [cpu activity].
//!
//! [Linux sysfs LED class]: https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-class-led
//! [`Led`]: trait.Led.html
//! [`RgbLed`]: trait.RgbLed.html
//! [`triggers`]: triggers/index.html
//! [timing]: triggers/trait.TriggerTimer.html
//! [heartbeat]: triggers/trait.TriggerHeartbeat.html
//! [cpu activity]: triggers/trait.TriggerCpu.html

// `error_chain!` can recurse deeply
#![recursion_limit = "1024"]

#[macro_use]
extern crate error_chain;

#[cfg(test)]
extern crate tempdir;

pub mod colors;
pub mod errors;
pub mod triggers;

use std::cmp;
use std::fs::OpenOptions;
use std::io::{Read, Write};
use std::path::{Path, PathBuf};

use colors::Color;
use errors::*;

const SYSFS_LED_CLASS: &'static str = "/sys/class/leds";


/// Brightness of an LED
///
/// Output brightness of an LED, always specified against some maximum
/// brightness. Usually this maximum brightness will be 255, but always prefer
/// the `Percent` variant over `Absolute` in case a value other than 255
/// exists.
///
/// ## Note
///
/// The Linux kernel defines the LED brightness as an enum type, so it could be
/// either 32 or 64 bits (or neither, the C spec is murky on this).
/// Realistically, however, it should never be more than 255 because that is
/// the definition for LED_FULL. We use u32 because it makes math easier.
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub enum Brightness {
    Full,
    Off,
    Percent(u32),
    Absolute(u32),
}

impl Brightness {
    pub fn to_absolute(&self, max_brightness: u32) -> u32 {
        match *self {
            Brightness::Full => max_brightness,
            Brightness::Off => 0,
            Brightness::Percent(p) => max_brightness.saturating_mul(cmp::min(p, 100)) / 100,
            Brightness::Absolute(a) => cmp::min(max_brightness, a),
        }
    }

    pub fn to_percent(&self, max_brightness: u32) -> u32 {
        match *self {
            Brightness::Full => max_brightness,
            Brightness::Off => 0,
            Brightness::Percent(p) => cmp::min(p, 100),
            Brightness::Absolute(a) => {
                cmp::min(a, max_brightness).saturating_mul(100) / max_brightness
            }
        }
    }
}

/// Basic functionality of an LED
///
/// Defines basic functionality of an LED, which is to be turned on or off at
/// some level of brightness.
pub trait Led {
    /// Get the current brightness of an LED
    fn brightness(&self) -> Result<Brightness>;
    /// Set the brightness of an LED
    fn set_brightness(&mut self, brightness: Brightness) -> Result<()>;
}

/// Access to an LED managed by the Linux LED sysfs class driver
pub struct SysfsLed {
    device_path: PathBuf,
}

impl SysfsLed {
    /// Create a new `SysfsLed` with a given name located in the default sysfs
    /// directory
    pub fn new(name: &str) -> Result<SysfsLed> {
        Self::from_path(Path::new(SYSFS_LED_CLASS).join(name))
    }

    /// Create a new `SysfsLed` with a custom path to the sysfs directory for
    /// the LED class device
    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<SysfsLed> {
        require_device_files(&path)?;
        Ok(SysfsLed { device_path: path.as_ref().to_path_buf() })
    }

    /// Return the raw max_brightness of the LED device
    pub fn max_brightness(&self) -> Result<u32> {
        Ok(self.sysfs_read_file("max_brightness")?.parse::<u32>()?)
    }

    fn sysfs_read_file(&self, name: &str) -> Result<String> {
        sysfs_read_file(&self.device_path, name)
    }

    fn sysfs_write_file(&self, name: &str, value: &str) -> Result<()> {
        sysfs_write_file(&self.device_path, name, value)
    }
}

impl Led for SysfsLed {
    fn brightness(&self) -> Result<Brightness> {
        Ok(Brightness::Absolute(self.sysfs_read_file("brightness")?.parse::<u32>()?))
    }

    fn set_brightness(&mut self, brightness: Brightness) -> Result<()> {
        let max_brightness = self.max_brightness()?;
        let string_value = format!("{}", brightness.to_absolute(max_brightness));
        self.sysfs_write_file("brightness", &string_value)?;
        Ok(())
    }
}

/// Basic functionality of an LED with red, green, and blue component colors
///
/// By stacking multiple LEDs together, one each of red, blue, and green, it is
/// possible to create a multicolored LED capable of showing many different
/// colors.
pub trait RgbLed: Led {
    /// Get the color of the RGB LED
    fn color(&self) -> Result<Color>;
    /// Set the color of the RGB LED
    fn set_color(&mut self, color: Color) -> Result<()>;
}

/// Access to an RGB LED managed by the Linux LED sysfs class driver,
/// configured as 3 separate LEDs.
pub struct SysfsRgbLed {
    red: SysfsLed,
    green: SysfsLed,
    blue: SysfsLed,
}

impl SysfsRgbLed {
    /// Create a new `SysfsRgbLed` from LEDs with the given names in the
    /// default sysfs directory
    pub fn new(red: &str, green: &str, blue: &str) -> Result<SysfsRgbLed> {
        Self::from_leds(SysfsLed::new(red)?,
                        SysfsLed::new(green)?,
                        SysfsLed::new(blue)?)
    }

    /// Create a new `SysfsRgbLed` with custom paths to the sysfs directories for
    /// the separate LED devices
    pub fn from_path<Pr, Pg, Pb>(red: Pr, green: Pg, blue: Pb) -> Result<SysfsRgbLed>
        where Pr: AsRef<Path>,
              Pg: AsRef<Path>,
              Pb: AsRef<Path>
    {
        Self::from_leds(SysfsLed::from_path(red)?,
                        SysfsLed::from_path(green)?,
                        SysfsLed::from_path(blue)?)
    }

    /// Create a new `SysfsRgbLed` from existing `SysfsLed` objects
    pub fn from_leds(red: SysfsLed, green: SysfsLed, blue: SysfsLed) -> Result<SysfsRgbLed> {
        Ok(SysfsRgbLed {
            red: red,
            green: green,
            blue: blue,
        })
    }
}

impl Led for SysfsRgbLed {
    // Brightness on an RGB LED as a whole is a bit strange since there are
    // three LEDs making up the output. We choose to treat brightness
    // "lightness" in the HSL color space instead - increasing lightness will
    // increase perceived brightness, so it's close.
    fn brightness(&self) -> Result<Brightness> {
        Ok(Brightness::Off)
    }

    fn set_brightness(&mut self, _brightness: Brightness) -> Result<()> {
        Ok(())
    }
}

impl RgbLed for SysfsRgbLed {
    fn color(&self) -> Result<Color> {
        // TODO: This isn't correct
        let _red = self.red.brightness()?;
        let _green = self.green.brightness()?;
        let _blue = self.blue.brightness()?;
        Ok(Color::from_rgb(0, 0, 0))
    }

    fn set_color(&mut self, color: Color) -> Result<()> {
        let red_max = self.red.max_brightness()? as u32;
        let green_max = self.green.max_brightness()? as u32;
        let blue_max = self.blue.max_brightness()? as u32;
        // TODO: This isn't correct
        self.red.set_brightness(Brightness::Absolute(color.red() as u32))?;
        self.green.set_brightness(Brightness::Absolute(color.green() as u32))?;
        self.blue.set_brightness(Brightness::Absolute(color.blue() as u32))?;
        Ok(())
    }
}

// Make sure that the specified files exist in the given directory
fn require_device_files<D>(dir: D) -> Result<()>
    where D: AsRef<Path>
{
    for file in &["brightness", "max_brightness", "trigger"] {
        if !dir.as_ref().join(file).is_file() {
            bail!(ErrorKind::InvalidDevicePath(dir.as_ref().to_string_lossy().into()));
        }
    }

    Ok(())
}

fn sysfs_read_file(device_path: &Path, name: &str) -> Result<String> {
    let path = device_path.join(name);
    let mut file = OpenOptions::new().read(true)
        .open(path)?;
    let mut result = String::new();
    file.read_to_string(&mut result)?;
    Ok(result.trim().into())
}

fn sysfs_write_file(device_path: &Path, name: &str, value: &str) -> Result<()> {
    let path = device_path.join(name);
    let mut file = OpenOptions::new().write(true)
        .truncate(true)
        .create(false)
        .open(path)?;
    Ok(file.write_all(value.as_bytes())?)
}

#[cfg(test)]
mod tests {
    use std::fs::File;
    use std::io::{Read, Write};
    use std::path::Path;

    use tempdir::TempDir;

    use super::*;

    struct SysfsWrapper(TempDir);

    impl SysfsWrapper {
        fn path(&self) -> &Path {
            self.0.path()
        }

        fn get(&self, name: &str) -> String {
            let mut result = String::new();
            File::open(self.path().join(name))
                .expect(&format!("opening {}", name))
                .read_to_string(&mut result)
                .expect(&format!("reading {}", name));
            result
        }

        fn set(&mut self, name: &str, value: &str) {
            File::open(self.path().join(name))
                .expect(&format!("opening {}", name))
                .write_all(value.as_bytes())
                .expect(&format!("writing {}", name));
        }
    }

    macro_rules! create_sysfs_dir {
        ( $name:expr; $( $file:expr => $value:expr );+ ) => {{
            let tempdir = TempDir::new($name).expect("create temp dir");
            $({
                let mut file = File::create(tempdir.path().join($file))
                    .expect(concat!("create ", $file, " file"));
                file.write_all($value.as_bytes())
                    .expect(concat!("writing ", $file, " initial value"));
            })+

            SysfsWrapper(tempdir)
        }};
    }

    #[test]
    fn test_set_brightness() {
        let harness = create_sysfs_dir!("sysfs_led_test";
                                        "brightness" => "0";
                                        "max_brightness" => "128";
                                        "trigger" => "[none]");
        let mut led = SysfsLed::from_path(harness.path()).expect("create sysfs led");
        let vectors = vec![(Brightness::Full, "128"),
                           (Brightness::Percent(50), "64"),
                           (Brightness::Percent(150), "128"),
                           (Brightness::Absolute(0), "0"),
                           (Brightness::Absolute(72), "72"),
                           (Brightness::Absolute(129), "128"),
                           (Brightness::Off, "0")];
        for (brightness, expected) in vectors {
            led.set_brightness(brightness).expect(&format!("setting brightness={:?}", brightness));
            assert_eq!(expected, harness.get("brightness"));
        }
    }
}
